homeassistant/components/octoprint/config_flow.py

```diff
@@ -0,0 +1,177 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _schema_with_defaults(username="", host=None, port=80, path="/", ssl=False):
+    return vol.Schema(
+        {
+            vol.Required(CONF_USERNAME, default=username): cv.string,
+            vol.Required(CONF_HOST, default=host): cv.string,
+            vol.Optional(CONF_PORT, default=port): cv.port,
+            vol.Optional(CONF_PATH, default=path): cv.string,
+            vol.Optional(CONF_SSL, default=ssl): cv.boolean,
+        },
+        extra=vol.ALLOW_EXTRA,
+    )
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for OctoPrint."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+    api_key_task = None
+
+    def __init__(self):
+        """Handle a config flow for OctoPrint."""
+        self.discovery_schema = None
+        self.user_input = {}
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        if user_input is None:
+            data = self.discovery_schema or _schema_with_defaults()
+            return self.async_show_form(step_id="user", data_schema=data)
+
+        self.user_input = user_input
+        if CONF_API_KEY in user_input:
+            return await self.async_step_finish(user_input)
+
+        self.api_key_task = None
+        return await self.async_step_get_api_key(user_input)
+
+    async def async_step_get_api_key(self, user_input=None):
+        """Get an Application Api Key."""
+        if not self.api_key_task:
+            self.api_key_task = self.hass.async_create_task(self._async_get_auth_key())
+            return self.async_show_progress(
+                step_id="get_api_key", progress_action="get_api_key"
+            )
+
+        try:
+            await self.api_key_task
+        except OctoprintException as err:
+            _LOGGER.error("Failed to get an application key : %s", err)
+            return self.async_show_progress_done(next_step_id="auth_failed")
+        except Exception as err:  # pylint: disable=broad-except
+            _LOGGER.error("Failed to get an application key : %s", err)
+            return self.async_show_progress_done(next_step_id="auth_failed")
+
+        return self.async_show_progress_done(next_step_id="finish")
+
+    async def async_step_finish(self, user_input=None):
+        """Finish the configuration setup."""
+        user_input = user_input or self.user_input
+        session = async_get_clientsession(self.hass)
+        octoprint = OctoprintClient(
+            user_input[CONF_HOST],
+            session,
+            user_input[CONF_PORT],
+            user_input[CONF_SSL],
+            user_input[CONF_PATH],
+        )
+        octoprint.set_api_key(user_input[CONF_API_KEY])
+
+        try:
+            discovery = await octoprint.get_discovery_info()
+        except ApiError as err:
+            _LOGGER.exception("Failed to connect to printer: %r", err)
```

Comment:
Known errors should be logged without stacktrace.
```suggestion
            _LOGGER.error("Failed to connect to printer: %r", err)
```